### PR TITLE
PR for Part 2 of the workshop

### PR DESCRIPTION
**Pull request by asteves**
This PR adds Part 2 to the workshop. It makes the following additions. 

- Structures the second part of the workshop around common data wrangling issues. These are based on Wickham (2014). 
- Reintroduces two-table verbs in dplyr and introduces `case_when()`
- Includes new datasets taken from real world data for students to practice with as well as show for general examples. All of the data files that are created are found in `dataCreation.R`
- Provides examples and challenges to integrate Part 1 material for students to practice with and as a reference 
- Introduces additional functions from `tidyr` (`expand()`, `nesting()`), `readr`, and `stringr`. The examples from `stringr` provide an extremely limited but functional introduction to regular expressions in the context of solving a data wrangling problem. 
- Adds four new challenges along with solutions. 

Applications to real world data wrangling in this PR are: 

- Creating a balanced panel/longitudinal dataset 
- Cleaning a survey crosstab
- Bulk recoding data via lookup tables, a problem that I (@asteves) originally solved during a D-Lab consulting project 
- Reading in multiple datasets with a common format at once.

### CODE REVIEW

File: lessons/Part2.Rmd [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+2. For each data frame, add a new column that records the original filename
+3. Combine all the data frames into a single data frame. 
+
+This pattern is so common that it is now built into the `readr` package. 
+
+```{r, eval = F}
+# Get appropriate files. 
+# Here we assume that we have .csv files, but any 
+# support file type will work
+files <- fs::dir_ls(path = "/PATH", glob = "*pattern*csv")
```
Comment (pssachdeva) [inline lessons/Part2.Rmd:376]: is there a simple example of this we could use to demonstrate the end result?

**Review by pssachdeva — CHANGES_REQUESTED** [review]
This is fantastic, thanks for all the work on this. I left some comments - most are simple fixes. A couple might require a little bit of reorganization. Feel free to ping me if you have any questions.

File: solutions/r_wrang_solutions.Rmd [hunk 2]
```diff
@@ -151,3 +151,98 @@ gapminder_long <- gapminder_wide %>%
   )
 ```
 
+## Challenge 9
```
[the remark this answers is not in the document]
Comment (asteves) [reply solutions/r_wrang_solutions.Rmd:154]: Done.

File: lessons/Part2.Rmd [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+
+1. Column headers that are values, not variable names. 
+2. Overloading variables into a single column. 
+3. Storing variables in both rows and columns. 
+4. Storing multiple types of units in the same data frame. 
+5. Storing a single type in multiple tables. 
+
+Within our discussion of (4) and (5), we will cover two table verbs in the tidyverse. In real data analysis projects, we will rarely have one data frame to do all of our analysis. 
+
+## Column headers that are values, not variable names.
```
[the remark this answers is not in the document]
Comment (asteves) [reply lessons/Part2.Rmd]: Done.

File: lessons/Part2.Rmd (hunk 1 above)
Comment (asteves) [reply lessons/Part2.Rmd:372]: The challenge is a demonstration of the end result. Do you mean in addition?

File: lessons/Part2.Rmd [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+```
+
+Once again, we get a tidy data frame that we can work with for future analyses. 
+
+
+## Storing multiple types of units in the same data frame. 
+
+A fundamental principle of tidy data, and data wrangling in general, is to keep like for like in the same table. Datasets often violate this guideline when they have values collected at multiple levels on different types of observations. 
+
+### Two Table Verbs
```
[the remark this answers is not in the document]
Comment (asteves) [reply lessons/Part2.Rmd:227]: Removed

File: lessons/Part2.Rmd (hunk 1 above)
Comment (pssachdeva) [reply lessons/Part2.Rmd:372]: I see, I missed that. Nevermind, disregard. Thanks for pointing that out.

File: lessons/Part2.Rmd [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+               names_to = c("observation", ".value"),
+               names_sep = "_",
+               values_to = "value")
+
+```
+
+Once again, we get a tidy data frame that we can work with for future analyses. 
+
+
+## Storing multiple types of units in the same data frame.
```
[the remark this answers is not in the document]
Comment (asteves) [reply lessons/Part2.Rmd]: Fixing the title I think resolves.

**Comment by asteves** [comment]
All comments and suggestion should now be resolved. Please let me know if it looks good to merge. @pssachdeva

**Comment by pssachdeva** [comment]
Awesome, give it a merge whenever you're ready!

**Pull request merged**